File: pipeline/src/ffrank/softsignals.py
"""Soft signals (spec §6b, §7, §8) — manual situational ratings, fully offline.

Situational judgment the stats can't see (QB, offensive line, scheme/coaching, pace, the player's
role, target/touch competition) is entered BY HAND on a 1-5 scale and turned into a per-player
`soft_score` multiplier: adjusted_points = base_points × soft_score. The board then re-ranks on
adjusted_points (ranking.ranking_metric already prefers it).

Two committed inputs under pipeline/data/ (edited in the app's Soft Signals studio, then saved):
  - team_situations.json  -> [{team, qb, ol, scheme, pace, notes}]   (team-wide factors, 1-5)
  - player_overrides.json -> [{id, role, competition}]               (player factors, 1-5)

Scale: 5 = best, 3 = neutral (no effect), 1 = worst. Each factor's deviation from 3 is weighted
by a position-specific table and summed; the result is clamped to [0.85, 1.15]. A player with all
factors neutral nets zero and is left unscored (ranks on base) — the LLM-off fallback equivalent.
"""
from __future__ import annotations

import json
from pathlib import Path

from .schema import Player, Situation

DATA_DIR = Path(__file__).resolve().parents[2] / "data"
TEAM_TABLE_PATH = DATA_DIR / "team_situations.json"
PLAYER_OVERRIDES_PATH = DATA_DIR / "player_overrides.json"

SOFT_MIN, SOFT_MAX = 0.85, 1.15
NEUTRAL = 3  # 1-5 scale; 3 = neutral / no effect.

# The six non-overlapping factors. Four are team-wide (one rating per team), two are per player.
TEAM_FACTORS = ("qb", "ol", "scheme", "pace")
PLAYER_FACTORS = ("role", "competition")
FACTORS = TEAM_FACTORS + PLAYER_FACTORS

# Human labels for the auto-generated reasoning string.
FACTOR_LABEL = {
    "qb": "QB", "ol": "OL", "scheme": "scheme", "pace": "pace",
    "role": "role", "competition": "competition",
}

# Position-specific weights — the max swing each factor can contribute. Each row SUMS TO 0.15,
# so "everything maxed" lands exactly on the ±15% bound (clamp is then only a safety net).
# Tunable: this table is the single source of truth (mirrored in app/lib/softsignals.ts).
FACTOR_WEIGHTS: dict[str, dict[str, float]] = {
    # qb is 0 for RB (their own passing doesn't matter); OL leads, role close behind.
    "RB": {"qb": 0.000, "ol": 0.050, "scheme": 0.025, "pace": 0.015, "role": 0.040, "competition": 0.020},
    # OL barely matters for a WR; QB and target competition lead.
    "WR": {"qb": 0.045, "ol": 0.000, "scheme": 0.030, "pace": 0.020, "role": 0.020, "competition": 0.035},
    "TE": {"qb": 0.040, "ol": 0.000, "scheme": 0.035, "pace": 0.020, "role": 0.035, "competition": 0.020},
    # For a QB the "qb" factor is themselves (0); competition = quality of weapons; OL = protection.
    "QB": {"qb": 0.000, "ol": 0.045, "scheme": 0.040, "pace": 0.030, "role": 0.015, "competition": 0.020},
}


def _clamp(x: float) -> float:
    return max(SOFT_MIN, min(SOFT_MAX, x))


# ----- Load the two committed rating files ----------------------------------------------

def load_team_ratings(path: Path = TEAM_TABLE_PATH) -> dict[str, dict]:
    """team abbrev -> {qb, ol, scheme, pace, notes}. Missing factors default to NEUTRAL."""
    table: dict[str, dict] = {}
    if Path(path).exists():
        for r in json.loads(Path(path).read_text()):
            team = r.get("team")
            if not team:
                continue
            row = {f: int(r.get(f, NEUTRAL)) for f in TEAM_FACTORS}
            row["notes"] = r.get("notes", "")
            table[team] = row
    return table


def load_player_overrides(path: Path = PLAYER_OVERRIDES_PATH) -> dict[str, dict]:
    """player id -> {role, competition}. Missing factors default to NEUTRAL."""
    overrides: dict[str, dict] = {}
    if Path(path).exists():
        for r in json.loads(Path(path).read_text()):
            pid = r.get("id")
            if not pid:
                continue
            overrides[pid] = {f: int(r.get(f, NEUTRAL)) for f in PLAYER_FACTORS}
    return overrides


def player_ratings(p: Player, team_ratings: dict[str, dict],
                   player_overrides: dict[str, dict]) -> dict[str, int]:
    """The six 1-5 ratings for a player: team factors from its team + its player overrides."""
    trow = team_ratings.get(p.team, {})
    orow = player_overrides.get(p.id, {})
    ratings = {f: int(trow.get(f, NEUTRAL)) for f in TEAM_FACTORS}
    ratings.update({f: int(orow.get(f, NEUTRAL)) for f in PLAYER_FACTORS})
    return ratings


# ----- The model -------------------------------------------------------------------------

def _reasoning(ratings: dict[str, int], weights: dict[str, float]) -> str:
    """One-line summary naming only the factors that actually move the score (weight > 0)."""
    lifts, drags = [], []
    for f in FACTORS:
        if ratings[f] == NEUTRAL or weights.get(f, 0.0) == 0.0:
            continue
        chip = f"{FACTOR_LABEL[f]} {ratings[f]}/5"
        (lifts if ratings[f] > NEUTRAL else drags).append(chip)
    parts = []
    if lifts:
        parts.append(", ".join(lifts) + " lift")
    if drags:
        parts.append(", ".join(drags) + " drag")
    return "; ".join(parts)


def soft_for_ratings(ratings: dict[str, int], position: str) -> float:
    """Position-weighted multiplier from the six ratings, clamped to [0.85, 1.15]."""
    weights = FACTOR_WEIGHTS.get(position, {})
    raw = sum(weights.get(f, 0.0) * (ratings[f] - NEUTRAL) / 2 for f in FACTORS)
    return round(_clamp(1.0 + raw), 4)


def compute_soft_scores(players: list[Player], team_ratings: dict[str, dict],
                        player_overrides: dict[str, dict]) -> dict[str, tuple[float, str]]:
    """id -> (soft_score, reasoning) for every player whose ratings have a NET effect.

    Players with all-neutral (or only zero-weight) ratings are omitted, so they stay unscored
    and rank on base_points — the same fallback behavior as 'no soft scores provided'."""
    scores: dict[str, tuple[float, str]] = {}
    for p in players:
        ratings = player_ratings(p, team_ratings, player_overrides)
        weights = FACTOR_WEIGHTS.get(p.position, {})
        raw = sum(weights.get(f, 0.0) * (ratings[f] - NEUTRAL) / 2 for f in FACTORS)
        if abs(raw) < 1e-9:
            continue  # no net effect -> leave on base
        scores[p.id] = (round(_clamp(1.0 + raw), 4), _reasoning(ratings, weights))
    return scores


# ----- Attach / apply --------------------------------------------------------------------

def attach_situation(players: list[Player], team_ratings: dict[str, dict],
                     player_overrides: dict[str, dict]) -> None:
    """Record each player's six ratings on situation.soft_factors (for the app/contract display).
    Players with all-neutral ratings are left untouched to keep the output clean."""
    for p in players:
        ratings = player_ratings(p, team_ratings, player_overrides)
        if all(v == NEUTRAL for v in ratings.values()):
            continue
        if p.situation is None:
            p.situation = Situation()
        p.situation.soft_factors = ratings


def apply_soft_scores(players: list[Player], scores: dict[str, tuple[float, str]]) -> int:
    """Set soft_score/reasoning for matched players. adjusted_points is computed later by
    vegas.finalize_adjusted (which composes soft × Vegas). Returns count matched."""
    n = 0
    for p in players:
        hit = scores.get(p.id)
        if hit is None:
            continue
        soft, reasoning = hit
        if p.situation is None:
            p.situation = Situation()
        p.situation.soft_score = soft
        p.situation.soft_reasoning = reasoning
        n += 1
    return n


def audit(players: list[Player], top: int = 15) -> list[Player]:
    """Biggest movers by |adjusted − base| (spec §9), for human review of outliers."""
    moved = [p for p in players if p.projection.adjusted_points is not None]
    moved.sort(key=lambda p: abs(p.projection.adjusted_points - p.projection.base_points), reverse=True)
    return moved[:top]

File: app/lib/softsignals.ts
import type { Player, PlayerOverride, TeamSituation } from "./types";

// Mirror of pipeline/src/ffrank/softsignals.py — keep the factor list, weights, and formula in sync.

export const SOFT_MIN = 0.85;
export const SOFT_MAX = 1.15;
export const NEUTRAL = 3; // 1-5 scale; 3 = neutral / no effect.
// Stay-put veterans get half their soft deviation (base_points already encodes their situation);
// rookies + team-changers (player.new_env) get full strength. Mirrors softsignals.py.
export const STAY_PUT_SOFT_WEIGHT = 0.5;

export const TEAM_FACTORS = ["qb", "ol", "scheme", "pace"] as const;
export const PLAYER_FACTORS = ["role", "competition"] as const;
export const FACTORS = [...TEAM_FACTORS, ...PLAYER_FACTORS] as const;
export type Factor = (typeof FACTORS)[number];

export const FACTOR_LABEL: Record<Factor, string> = {
  qb: "QB",
  ol: "OL",
  scheme: "Scheme",
  pace: "Pace",
  role: "Role",
  competition: "Competition",
};

// One-line explainers behind the ⓘ. Rate the CHANGE vs last year — the projection already
// reflects last season's situation, so 3 = no change = no effect, 5 = clear upgrade, 1 = decline.
export const FACTOR_INFO: Record<Factor, string> = {
  qb: "Change in QB play vs last year. 5 = clear upgrade, 1 = clear downgrade, 3 = same QB → no effect. Biggest lever for WR/TE; little for RBs; ignored for a QB rating his own team.",
  ol: "Change in the offensive line vs last year. 5 = materially better (key additions), 1 = materially worse (losses), 3 = unchanged. Biggest lever for RBs; negligible for WR/TE.",
  scheme: "Change in scheme / play-caller vs last year. 5 = new system that fits this position better, 1 = worse fit, 3 = same scheme.",
  pace: "Change in the offense's volume or pass-run lean vs last year. 5 = more favorable for this position, 1 = less favorable, 3 = unchanged.",
  role: "Change in this player's role/opportunity vs last year. 5 = bigger role (promotion, vacated touches), 1 = smaller role (new committee, lost a job), 3 = unchanged. Rookies: rate the role they're stepping into.",
  competition: "Change in target/touch competition vs last year. 5 = competition left (cleaner path), 1 = competition added (draft pick/signing), 3 = unchanged. Rookies: rate the room they land in.",
};

// Each position's weights SUM TO 0.15 → "everything maxed" lands exactly on the ±15% bound.
export const WEIGHTS: Record<string, Record<Factor, number>> = {
  RB: { qb: 0.0, ol: 0.05, scheme: 0.025, pace: 0.015, role: 0.04, competition: 0.02 },
  WR: { qb: 0.045, ol: 0.0, scheme: 0.03, pace: 0.02, role: 0.02, competition: 0.035 },
  TE: { qb: 0.04, ol: 0.0, scheme: 0.035, pace: 0.02, role: 0.035, competition: 0.02 },
  QB: { qb: 0.0, ol: 0.045, scheme: 0.04, pace: 0.03, role: 0.015, competition: 0.02 },
};

const clamp = (x: number) => Math.max(SOFT_MIN, Math.min(SOFT_MAX, x));
const round4 = (x: number) => Math.round(x * 1e4) / 1e4;

export type Ratings = Record<Factor, number>;

/** The six 1-5 ratings for a player: team factors from its team + its player overrides. */
export function playerRatings(
  player: Player,
  teamRatings: Record<string, TeamSituation>,
  overrides: Record<string, PlayerOverride>,
): Ratings {
  const t = teamRatings[player.team] ?? {};
  const o = overrides[player.id] ?? {};
  return {
    qb: t.qb ?? NEUTRAL,
    ol: t.ol ?? NEUTRAL,
    scheme: t.scheme ?? NEUTRAL,
    pace: t.pace ?? NEUTRAL,
    role: o.role ?? NEUTRAL,
    competition: o.competition ?? NEUTRAL,
  };
}

function reasoning(ratings: Ratings, weights: Record<Factor, number>): string {
  const lifts: string[] = [];
  const drags: string[] = [];
  for (const f of FACTORS) {
    if (ratings[f] === NEUTRAL || (weights[f] ?? 0) === 0) continue;
    const chip = `${FACTOR_LABEL[f]} ${ratings[f]}/5`;
    (ratings[f] > NEUTRAL ? lifts : drags).push(chip);
  }
  const parts: string[] = [];
  if (lifts.length) parts.push(lifts.join(", ") + " lift");
  if (drags.length) parts.push(drags.join(", ") + " drag");
  return parts.join("; ");
}

export interface SoftResult {
  soft: number;
  reasoning: string;
  effect: boolean; // false when ratings net to no change (player ranks on base)
  shrunk: boolean; // true when a stay-put vet's deviation was halved
  ratings: Ratings;
}

/** Position-weighted multiplier from a player's ratings, scaled by how blind the base is. */
export function computeSoft(
  player: Player,
  teamRatings: Record<string, TeamSituation>,
  overrides: Record<string, PlayerOverride>,
): SoftResult {
  const ratings = playerRatings(player, teamRatings, overrides);
  const weights = WEIGHTS[player.position] ?? ({} as Record<Factor, number>);
  let raw = FACTORS.reduce((s, f) => s + (weights[f] ?? 0) * ((ratings[f] - NEUTRAL) / 2), 0);
  const effect = Math.abs(raw) >= 1e-9;
  const shrunk = effect && !player.new_env;
  if (shrunk) raw *= STAY_PUT_SOFT_WEIGHT; // base already knows a stay-put vet → trust the rating half as much
  return { soft: round4(clamp(1 + raw)), reasoning: reasoning(ratings, weights), effect, shrunk, ratings };
}

function download(filename: string, body: unknown) {
  const blob = new Blob([JSON.stringify(body, null, 2)], { type: "application/json" });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  a.click();
  URL.revokeObjectURL(url);
}

/** Emit pipeline/data/team_situations.json — only teams with a non-neutral factor or a note. */
export function downloadTeams(teamRatings: Record<string, TeamSituation>) {
  const rows = Object.entries(teamRatings)
    .map(([team, s]) => ({
      team,
      qb: s.qb ?? NEUTRAL,
      ol: s.ol ?? NEUTRAL,
      scheme: s.scheme ?? NEUTRAL,
      pace: s.pace ?? NEUTRAL,
      notes: s.notes ?? "",
    }))
    .filter((r) => r.qb !== NEUTRAL || r.ol !== NEUTRAL || r.scheme !== NEUTRAL || r.pace !== NEUTRAL || r.notes)
    .sort((a, b) => a.team.localeCompare(b.team));
  download("team_situations.json", rows);
}

/** Emit pipeline/data/player_overrides.json — only players with a non-neutral factor. */
export function downloadOverrides(overrides: Record<string, PlayerOverride>) {
  const rows = Object.values(overrides)
    .map((o) => ({ id: o.id, role: o.role ?? NEUTRAL, competition: o.competition ?? NEUTRAL }))
    .filter((r) => r.role !== NEUTRAL || r.competition !== NEUTRAL);
  download("player_overrides.json", rows);
}
